### ISSUE TITLE
Questions about head (fc7) - dimensions and annotations

### ISSUE THREAD

**Issue by AnshKhurana** [shelhamer/revolver#6]
#### classification head (including fc7 for compatibility with guidance)
        head = [('fc7', nn.Conv2d(self.feat_dim*2, self.feat_dim*2, 1)),
            ('fc7_relu', nn.ReLU(inplace=True)),
            ('fc7_drop', nn.Dropout2d(p=0.5)),
            ('score', nn.Conv2d(self.feat_dim*2, 1, 1))]
        self.head = nn.Sequential(OrderedDict(head))

Here, why are the dimensions hard-coded to feat_dim * 2? Considering the positive map m(+), negative map m(-) annotations which are concatenated with im_feats, shouldn't this number be feat_dim*3?

#### align image + mask, then mask features by annotations for guidance
        im_feats = self.mask_feat(im, annos[0], scale=False)
        guides = [self.mask_feat(im, a) for a in annos]
#### stack image-guidance pairs into batch dimension
        feat = torch.cat([torch.cat((im_feats, g), dim=1) for g in guides], dim=0)

Here, it seems as if the code allows multiple guidance channels to be concatenated with the image features, which further adds to the confusion since the dimensions are fixed to be feat_dim*2.

In general, it'll be great if one can shed some light on the incorporation of annos/guides in the architecture.

Thank you in advance.

**Comment by shelhamer** [comment]
Note that the "late" models learn a metric that is **shared** across classes. The same weights define the distance between an image feature and guidance for each class, meaning that positives and negatives share the same metric, calculated for the positive and negative guidance features.

This is why the head has 2x the feature dimension: it is the concatenation of input features and guidance features. This is why the head has only one output: it is the score for that input-guidance pair.

> Here, it seems as if the code allows multiple guidance channels to be concatenated with the image features [...]

Correct. This is why the metric weights are shared: by defining the weights on input-guidance **pairs**, we can learn a single model that takes different numbers of guidance channels. Note the "stack image-guidance pairs into the batch dimension" comment, which points out that each input-guidance concatenation is packed into the batch dimension. Putting each pair along the batch dimension makes them share the same weights for the metric.

Hope that helps in making sense of the code.

The exception to this is the [unshared model](https://github.com/shelhamer/revolver/blob/master/revolver/model/cofeat_late_lite_unshared.py), which learns different weights for the positives and negatives.
